### PR TITLE
Remove LTP sudo, Change sudo as parameter, Add full test support for LTP

### PR DESCRIPTION
**Pull request by paxue**
Add full test support for LTP
Remove LTP sudo, change as parameter

### CODE REVIEW

File: microsoft/testsuites/ltp/ltp.py [hunk 1]
```diff
@@ -72,11 +72,12 @@ def dependencies(self) -> List[Type[Tool]]:
     def can_install(self) -> bool:
         return True
 
-    def __init__(self, node: Node, source_file: str, *args: Any, **kwargs: Any) -> None:
-        super().__init__(node, args, kwargs)
+    def __init__(self, node: Node, *args: Any, **kwargs: Any) -> None:
```
Comment (squirrelsc) [inline microsoft/testsuites/ltp/ltp.py]: Give the value with a default value, instead of using kwargs.get

File: microsoft/testsuites/ltp/ltp.py [hunk 2]
```diff
@@ -72,11 +72,12 @@ def dependencies(self) -> List[Type[Tool]]:
...
+        super().__init__(node, *args, **kwargs)
         git_tag = kwargs.get("git_tag", "")
         self._git_tag = git_tag if git_tag else self.DEFAULT_LTP_TESTS_GIT_TAG
-        self._source_file = source_file
+        self._source_file = kwargs.get("source_file", "")
+        self._run_as_sudo = kwargs.get("sudo", True)
```
Comment (squirrelsc) [inline microsoft/testsuites/ltp/ltp.py]: A similar question like Kselftest. Any difference to run with/out sudo?

File: microsoft/testsuites/ltp/ltp.py [hunk 3]
```diff
@@ -87,6 +88,9 @@ def run_test(
         block_device: Optional[str] = None,
         temp_dir: str = "/tmp/",
         ltp_run_timeout: int = 12000,
+        # run full test if True, else run lite test
+        run_full_test: bool = False,
```
Comment (squirrelsc) [inline microsoft/testsuites/ltp/ltp.py:92]: If the list is not given, it can run full. So it doesn't need the additional parameter.

File: microsoft/testsuites/ltp/ltpsuite.py [hunk 4]
```diff
@@ -71,6 +71,9 @@ def verify_ltp_lite(
         # If not provided, we will find a disk with enough space
         block_device = variables.get("ltp_block_device", None)
 
+        run_full_ltp_test = variables.get("run_full_ltp_test", False)
```
Comment (squirrelsc) [inline microsoft/testsuites/ltp/ltpsuite.py]: Is there any requirement to run all of them? This flag is easy to be conflict with other parameters. Please create another test case called verify_ltp. The only difference is the new test case is to run all, if no list given.

File: microsoft/testsuites/ltp/ltp.py (hunk 1 above)
Comment (paxue) [reply microsoft/testsuites/ltp/ltp.py]: Do you mean add the following lines in __init__()?
   self._source_file = ""
   self._run_as_sudo = True

Comment (squirrelsc) [reply microsoft/testsuites/ltp/ltp.py]: it should be like `def __init__(self, node: Node, source_file: str = "", run_as_sudo:bool =True, *args: Any, **kwargs: Any) -> None:`

File: microsoft/testsuites/ltp/ltp.py (hunk 2 above)
Comment (paxue) [reply microsoft/testsuites/ltp/ltp.py]: For using released file.
Kselftest: don't have config, can direct unzip the release file and run.
LTP: does have config step, this step involve lots of sudo issue.

File: microsoft/testsuites/ltp/ltpsuite.py (hunk 4 above)
Comment (paxue) [reply microsoft/testsuites/ltp/ltpsuite.py]: Is there any requirement to run all of them?
[panfeng xue]: yes. it's required from Cooperate team

I am fine with 2 paths: Create another verify_ltp_full() or just using the same test suite verify_ltp_lite.
From my view, their changes are very limited, only one line code change. Depends how you think this question

File: microsoft/testsuites/ltp/ltp.py (hunk 2 above)
Comment (squirrelsc) [reply microsoft/testsuites/ltp/ltp.py]: I mean for LTP, if run without sudo, does it work?

File: microsoft/testsuites/ltp/ltpsuite.py (hunk 4 above)
Comment (squirrelsc) [reply microsoft/testsuites/ltp/ltpsuite.py]: I don't want to add a parameter called `run_full_ltp_test`, it conflicts with existing parameters like `ltp_test` and `ltp_skip_test`

File: lisa/operating_system.py [hunk 5]
```diff
@@ -388,9 +388,10 @@ def install_packages(
         signed: bool = False,
         timeout: int = 1200,
         extra_args: Optional[List[str]] = None,
+        sudo: bool = True,
```
Comment (squirrelsc) [inline lisa/operating_system.py]: did you find any package installation command are blocked? Please provide examples here. We can think about if it can be detected in code, instead of depending on configuration.

File: lisa/tools/pgrep.py [hunk 6]
```diff
@@ -42,10 +42,12 @@ def _freebsd_tool(cls) -> Optional[Type[Tool]]:
...
         return False
 
-    def get_processes(self, process_identifier: str) -> List[ProcessInfo]:
+    def get_processes(
+        self, process_identifier: str, sudo: bool = True
+    ) -> List[ProcessInfo]:
         running_process: List[ProcessInfo] = []
         output = self.run(
-            f'-l "{process_identifier}"', sudo=True, force_run=True
+            f'-l "{process_identifier}"', sudo=sudo, force_run=True
```
Comment (squirrelsc) [inline lisa/tools/pgrep.py]: The same above, what's the error message, when this command failed with sudo?

File: lisa/operating_system.py [hunk 7]
```diff
@@ -1922,7 +1922,9 @@ def __init__(self, node: Any) -> None:
         self._dnf_tool_name: str
 
     def _initialize_package_installation(self) -> None:
-        self.set_kill_user_processes()
+        self._log.debug(f"--- CBLMariner, type_name: {self._node.type_name()}")
```
Comment (squirrelsc) [inline lisa/operating_system.py]: clean up the temp log.

File: lisa/operating_system.py [hunk 8]
```diff
@@ -1922,7 +1922,9 @@ def __init__(self, node: Any) -> None:
...
+        if self._node.type_name() != "local":
```
Comment (squirrelsc) [inline lisa/operating_system.py]: 1. what does it need to skip local? if any reason, write it in the code comment.
2. use `isinstance to check types, not use a string`.

File: microsoft/testsuites/ltp/ltp.py [hunk 9]
```diff
@@ -87,14 +93,18 @@ def run_test(
...
     ) -> List[LtpResult]:
-        # tests cannot be empty
-        assert_that(ltp_tests, "ltp_tests cannot be empty").is_not_empty()
+        if not run_full_test:
+            # ltp_tests cannot be empty
+            assert_that(ltp_tests, "ltp_tests cannot be empty").is_not_empty()
```
Comment (squirrelsc) [inline microsoft/testsuites/ltp/ltp.py]: How about empty ltp_tests means run all? So it doesn't need the `fun_full_test` variable.

File: microsoft/testsuites/ltp/ltp.py [hunk 10]
```diff
@@ -327,12 +341,15 @@ def _install(self) -> bool:
...
         sysctl.write("vm.dirty_background_ratio", "5")
-        sysctl.run("-p")
+        sysctl.run("-p", sudo=True)
 
         # find partition to install ltp
-        build_dir = self.node.find_partition_with_freespace(
-            self.BUILD_REQUIRED_DISK_SIZE_IN_GB
-        )
+        if self.node.type_name() == "local":
+            build_dir = "/"
```
Comment (squirrelsc) [inline microsoft/testsuites/ltp/ltp.py]: Why use root for the local node?

File: microsoft/testsuites/ltp/ltpsuite.py [hunk 11]
```diff
@@ -105,6 +107,69 @@ def verify_ltp_lite(
...
+        ),
+    )
+    def verify_ltp_full(
+        self,
+        node: Node,
+        log_path: str,
+        variables: Dict[str, Any],
+        result: TestResult,
+    ) -> None:
+        run_as_sudo = variables.get("run_as_sudo", True)
```
Comment (squirrelsc) [inline microsoft/testsuites/ltp/ltpsuite.py]: remove `run_as_sudo`

File: microsoft/testsuites/ltp/ltp.py [hunk 12]
```diff
@@ -72,29 +72,33 @@ def dependencies(self) -> List[Type[Tool]]:
...
+        super().__init__(node, *args, **kwargs)
         git_tag = kwargs.get("git_tag", "")
         self._git_tag = git_tag if git_tag else self.DEFAULT_LTP_TESTS_GIT_TAG
         self._source_file = source_file
 
     def run_test(
         self,
         test_result: TestResult,
-        ltp_tests: List[str],
+        ltp_tests: List[str],  # empty will run full test
```
Comment (squirrelsc) [inline microsoft/testsuites/ltp/ltp.py]: Write comments above of code.

File: microsoft/testsuites/ltp/ltp.py [hunk 13]
```diff
@@ -72,29 +72,33 @@ def dependencies(self) -> List[Type[Tool]]:
...
         ltp_run_timeout: int = 12000,
     ) -> List[LtpResult]:
-        # tests cannot be empty
-        assert_that(ltp_tests, "ltp_tests cannot be empty").is_not_empty()
         ls = self.node.tools[Ls]
         rm = self.node.tools[Rm]
 
         # remove skipfile if it exists
-        if ls.path_exists(self.LTP_SKIP_FILE):
+        if ls.path_exists(self.LTP_SKIP_FILE, sudo=True):
```
Comment (squirrelsc) [inline microsoft/testsuites/ltp/ltp.py]: Is the sudo needed?

File: microsoft/testsuites/ltp/ltpsuite.py [hunk 14]
```diff
@@ -83,7 +83,7 @@ def verify_ltp_lite(
         else:
             skip_test_list = []
 
-        if not block_device:
+        if not block_device and node.type_name() != "local":
```
Comment (squirrelsc) [inline microsoft/testsuites/ltp/ltpsuite.py]: 1. Why not support local? Add code comments.
2. Use isinstance to check by type, not by a string.

File: microsoft/testsuites/ltp/ltpsuite.py [hunk 15]
```diff
@@ -105,6 +107,66 @@ def verify_ltp_lite(
...
+        ),
+    )
+    def verify_ltp_full(
+        self,
+        node: Node,
+        log_path: str,
+        variables: Dict[str, Any],
+        result: TestResult,
+    ) -> None:
+        ltp_source_file = variables.get("ltp_source_file", "")
```
Comment (squirrelsc) [inline microsoft/testsuites/ltp/ltpsuite.py]: As the new logic, it can share much more logic.

File: microsoft/testsuites/ltp/ltp.py [hunk 16]
```diff
@@ -111,8 +116,10 @@ def run_test(
         # add parameters for the test logging
         parameters = f"-p -q -l {self.LTP_RESULT_PATH} -o {self.LTP_OUTPUT_PATH} "
 
-        # add the list of tests to run
-        parameters += f"-f {','.join(ltp_tests)} "
+        # add the list of tests to run, only when ltp_tests is not empty
+        # if ltp_tests=[], "-f" not set, ltp will run full test
+        if len(ltp_tests) > 0:
```
Comment (squirrelsc) [inline microsoft/testsuites/ltp/ltp.py]: `if ltp_tests:` is enough.

File: microsoft/testsuites/ltp/ltp.py [hunk 17]
```diff
@@ -347,15 +356,19 @@ def _install(self) -> bool:
...
                 f"Use downloaded source code tar file: {self._source_file}!"
             )
+            remote_source_folder = self.get_tool_path(use_global=True)
             remote_source_file = self.get_tool_path(use_global=True) / os.path.basename(
                 self._source_file
             )
 
             copy = self.node.tools[RemoteCopy]
-            copy.copy_to_remote(PurePath(self._source_file), remote_source_file)
+            copy.copy_to_remote(
```
Comment (squirrelsc) [inline microsoft/testsuites/ltp/ltp.py]: Does it need sudo false?

File: microsoft/testsuites/ltp/ltp.py [hunk 18]
```diff
@@ -347,15 +356,19 @@ def _install(self) -> bool:
...
+                PurePath(self._source_file), remote_source_folder, sudo=False
+            )
             self.node.tools[Tar].extract(
-                str(remote_source_file), top_src_dir, sudo=True
+                str(remote_source_file), top_src_dir, strip_components=1
```
Comment (squirrelsc) [inline microsoft/testsuites/ltp/ltp.py]: What's the strip_components for? Why it can pass before?

File: microsoft/testsuites/ltp/ltpsuite.py [hunk 19]
```diff
@@ -32,6 +32,44 @@ class LtpTestsuite(TestSuite):
...
     LTP_REQUIRED_DISK_SIZE_IN_GB = 2
 
+    def _get_parameter(
+        self,
+        variables: Dict[str, Any],
+    ) -> None:
+        # read parameters from .yml file
+        tests = variables.get("ltp_test", "")
+        skip_tests = variables.get("ltp_skip_test", "")
+        self.ltp_source_file = variables.get("ltp_source_file", "")
```
Comment (squirrelsc) [inline microsoft/testsuites/ltp/ltpsuite.py]: Don't use class members in the test suite, it breaks the design. Use a top-level common method, and no output parameters needed.

File: microsoft/testsuites/ltp/ltpsuite.py [hunk 20]
```diff
@@ -19,6 +19,59 @@
...
+    binary_file = variables.get("ltp_binary_file", "")
+    # block device is required for few ltp tests
+    # If not provided, we will find a disk with enough space
+    block_device = variables.get("ltp_block_device", None)
+
+    # get comma separated list of tests
+    if test_str:
+        test_list = test_str.split(",")
+    else:
+        test_list = LTP_LITE_TESTS  # fallback lite tests
```
Comment (squirrelsc) [inline microsoft/testsuites/ltp/ltpsuite.py]: This is the only different code between lite and full test cases. All other logic should be shared. It doesn't need to create a test config class. Refer to smoke_test, how to use one method to serve multiple test cases.

File: microsoft/testsuites/ltp/ltpsuite.py [hunk 21]
```diff
@@ -32,57 +31,32 @@ class LtpTestsuite(TestSuite):
...
         self,
         node: Node,
-        log_path: str,
         variables: Dict[str, Any],
+        run_full_test: bool,
+        log_path: str,
         result: TestResult,
     ) -> None:
-        # parse variables
+        test_str = variables.get("ltp_test", "")
```
Comment (squirrelsc) [inline microsoft/testsuites/ltp/ltpsuite.py]: This line can be moved to the test case level, like below. It would be cleared on the difference.

```python
# verify_ltp_lite
test_str = variables.get("ltp_test", self.LTP_LITE_TESTS)

# verify_ltp_full
test_str = variables.get("ltp_test", "")
```

File: microsoft/testsuites/ltp/ltpsuite.py [hunk 22]
```diff
@@ -93,22 +61,92 @@ def verify_ltp_lite(
                     node.tools[Lsblk].find_disk_by_mountpoint(mountpoint).device_name
                 )
 
-        # run ltp lite tests
+        if not test_list:
+            # empty test_list means full test, require long timeout
+            run_timeout = 72000
```
Comment (squirrelsc) [inline microsoft/testsuites/ltp/ltpsuite.py]: It won't work, because the test case metadata has another timeout. This value couldn't be longer than that. 72000 means  20 hours, it's too long to finish all cases.

File: microsoft/testsuites/ltp/ltpsuite.py [hunk 23]
```diff
@@ -93,22 +61,92 @@ def verify_ltp_lite(
...
+        ),
+    )
+    def verify_ltp_lite(
+        self,
+        node: Node,
+        log_path: str,
+        variables: Dict[str, Any],
+        result: TestResult,
+    ) -> None:
+        test_str = variables.get("ltp_test", self.LTP_LITE_TESTS)
```
Comment (squirrelsc) [inline microsoft/testsuites/ltp/ltpsuite.py:66]: Change it like below, and change the argument type to string

```python
        test_str = variables.get("ltp_test", self.LTP_LITE_TESTS)
        self._verify_ltp(node, variables, test_str, log_path, result)
```

File: microsoft/testsuites/ltp/ltpsuite.py [hunk 24]
```diff
@@ -93,22 +61,92 @@ def verify_ltp_lite(
...
+        ),
+    )
+    def verify_ltp_full(
+        self,
+        node: Node,
+        log_path: str,
+        variables: Dict[str, Any],
+        result: TestResult,
+    ) -> None:
+        self._verify_ltp(node, variables, [], log_path, result)
```
Comment (squirrelsc) [inline microsoft/testsuites/ltp/ltpsuite.py]: Change it like below, and change the argument type to string

```python
        test_str = variables.get("ltp_test", "")
        self._verify_ltp(node, variables, test_str, log_path, result)
```

File: microsoft/testsuites/ltp/ltpsuite.py [hunk 25]
```diff
@@ -29,60 +28,29 @@
...
-            min_core_count=8,
-            disk=schema.DiskOptionSettings(
-                data_disk_count=search_space.IntRange(min=1),
-                data_disk_size=search_space.IntRange(min=12),
-            ),
-            unsupported_os=[BSD, Windows],
-        ),
-    )
-    def verify_ltp_lite(
+    def _verify_ltp(
```
Comment (squirrelsc) [inline microsoft/testsuites/ltp/ltpsuite.py]: In a class, move the private method below public methods.

**Comment by squirrelsc** [comment]
@LiliDeng LGTM

**Comment by LiliDeng** [comment]
@paxue could you please rebase the latest main, i saw permission issue on mariner against this PR.

**Review by LiliDeng — APPROVED** [review]
